flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheCleanupTest.java

```diff
@@ -0,0 +1,324 @@
+				while (checkFilesExist(jobId, keys, cache, false) != 0 &&
+					System.currentTimeMillis() < deadline);
+			}
+
+			// the blob cache should no longer contain the files
+			// this fails if we exited via a timeout
+			checkFileCountForJob(0, jobId, cache);
+			// server should be unaffected
+			checkFileCountForJob(2, jobId, server);
+		}
+		finally {
+			if (cache != null) {
+				cache.close();
+			}
+
+			if (server != null) {
+				server.close();
+			}
+			// now everything should be cleaned up
+			checkFileCountForJob(0, jobId, server);
+		}
+	}
+
+	/**
+	 * Tests that {@link BlobCache} cleans up after calling {@link BlobCache#releaseJob(JobID)}
+	 * but only after preserving the file for a bit longer.
+	 */
+	@Test
+	@Ignore("manual test due to stalling: ensures a BLOB is retained first and only deleted after the (long) timeout ")
+	public void testJobDeferredCleanup() throws IOException, InterruptedException {
+		// file should be deleted between 5 and 10s after last job release
+		long cleanupInterval = 5L;
+
+		JobID jobId = new JobID();
+		List<BlobKey> keys = new ArrayList<BlobKey>();
+		BlobServer server = null;
+		BlobCache cache = null;
+
+		final byte[] buf = new byte[128];
+
+		try {
+			Configuration config = new Configuration();
+			config.setString(BlobServerOptions.STORAGE_DIRECTORY,
+				temporaryFolder.newFolder().getAbsolutePath());
+			config.setLong(BlobServerOptions.CLEANUP_INTERVAL, cleanupInterval);
+
+			server = new BlobServer(config, new VoidBlobStore());
+			InetSocketAddress serverAddress = new InetSocketAddress("localhost", server.getPort());
+			BlobClient bc = new BlobClient(serverAddress, config);
+			cache = new BlobCache(serverAddress, config, new VoidBlobStore());
+
+			keys.add(bc.put(jobId, buf));
+			buf[0] += 1;
+			keys.add(bc.put(jobId, buf));
+
+			bc.close();
+
+			checkFileCountForJob(2, jobId, server);
+			checkFileCountForJob(0, jobId, cache);
+
+			// register once
+			cache.registerJob(jobId);
+
+			checkFileCountForJob(2, jobId, server);
+			checkFileCountForJob(0, jobId, cache);
+
+			for (BlobKey key : keys) {
+				cache.getFile(jobId, key);
+			}
+
+			// register again (let's say, from another thread or so)
+			cache.registerJob(jobId);
+			for (BlobKey key : keys) {
+				cache.getFile(jobId, key);
+			}
+
+			assertEquals(2, checkFilesExist(jobId, keys, cache, true));
+			checkFileCountForJob(2, jobId, server);
+			checkFileCountForJob(2, jobId, cache);
+
+			// after releasing once, nothing should change
+			cache.releaseJob(jobId);
+
+			assertEquals(2, checkFilesExist(jobId, keys, cache, true));
+			checkFileCountForJob(2, jobId, server);
+			checkFileCountForJob(2, jobId, cache);
+
+			// after releasing the second time, the job is up for deferred cleanup
+			cache.releaseJob(jobId);
+
+			// files should still be accessible for now
+			assertEquals(2, checkFilesExist(jobId, keys, cache, true));
+			checkFileCountForJob(2, jobId, cache);
+
+			Thread.sleep(cleanupInterval / 5);
```

Comment:
It's hard to test a functionality based on time that way though...

For exactly this reason (and the long waiting time), this test is ignored and should be manually tested instead.